File: Graphs/994_rotting_oranges/994.rs
impl Solution {
    pub fn oranges_rotting(grid: Vec<Vec<i32>>) -> i32 {
        
        let rows: usize = grid.len();
        let cols: usize = grid[0].len();
    
        let directions: Vec<(isize, isize)> = vec!((-1, 0), (0, 1), (1, 0), (0, -1));
        
        fn rotting_process(grid: &mut Vec<Vec<i32>>, rows: &usize, cols: &usize, directions: &Vec<(isize, isize)>, timestamp: i32) -> bool {
            let mut continue_process: bool = false;
            for row in 0..*rows {
                for col in 0..*cols {
                    if grid[row][col] == timestamp {
                        for (mut r, mut c) in directions.iter() {
                            r += row as isize;
                            c += col as isize;
                            if *rows as isize > r && r >= 0 && *cols as isize > c && c >= 0 && grid[r as usize][c as usize] == 1 {
                                grid[r as usize][c as usize] = timestamp + 1;
                                continue_process = true;
                            }
                        }
                    }
                }
            }
            return continue_process;
        }
        
        let mut timestamp: i32 = 2;
        let mut grid = grid;
        while rotting_process(&mut grid, &rows, &cols, &directions, timestamp) {
            timestamp += 1;
        }
        
        for row in grid.iter() {
            for cell in row.iter() {
                if *cell == 1 {
                    return -1;
                }
            }
        }
        
        return timestamp - 2;
    
    }
}

File: Graphs/994_rotting_oranges/994.py
class Solution:
    def orangesRotting(self, grid) -> int:
        
        row_len = len(grid)
        col_len = len(grid[0])
        
        queue = []
        minutes = set()
        
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        
        def DFS(queue):
            
            minutes.clear()

            while queue:
                row, col, minute = queue.pop(0)
                minutes.add(minute)
                for dirct in directions:
                    r = row + dirct[0]
                    c = col + dirct[1]

                    if r >= 0 and r < row_len and c >= 0 and c < col_len and grid[r][c] == 1:
                        queue.append([r, c, minute + 1])
                        grid[r][c] = 0
            
            if len(minutes) == 0:
                return 0
            return max(minutes)
                
        
        for i in range(row_len):
            for j in range(col_len):
                if grid[i][j] == 2:
                    queue.append([i, j, 0])
                    
        time = DFS(queue)
        
        for rows in grid:
            if 1 in rows:
                return -1
        
        return time